File: src/main/java/com/braydenoneal/blang/parser/expression/value/BlockValue.java
package com.braydenoneal.blang.parser.expression.value;

import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.block.Block;
import net.minecraft.registry.Registries;

public class BlockValue extends Value<Block> {
    public static final MapCodec<BlockValue> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
            Registries.BLOCK.getCodec().fieldOf("value").forGetter(BlockValue::value)
    ).apply(instance, BlockValue::new));

    public BlockValue(Block value) {
        super(value);
    }

    @Override
    public ValueType<?> getType() {
        return ValueTypes.BLOCK;
    }

    @Override
    public String toString() {
        return "\"" + value() + "\"";
    }

    @Override
    public boolean equals(Object obj) {
        try {
            if (obj instanceof BlockValue blockValue) {
                return value().equals(blockValue.value());
            }
        } catch (Error e) {
            System.out.println("Cannot equate block values outside of the game");
        }

        return false;
    }
}

File: src/main/kotlin/com/braydenoneal/blang/parser/expression/value/BlockValue.kt
package com.braydenoneal.blang.parser.expression.value

import com.braydenoneal.blang.parser.RunException
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.block.Block
import net.minecraft.registry.Registries

class BlockValue(value: Block) : Value<Block>(value) {
    override val valueType: ValueType<*> get() = ValueTypes.BLOCK

    override fun equals(other: Any?): Boolean {
        try {
            if (other is BlockValue) {
                return value == other.value
            }
        } catch (_: Error) {
            throw RunException("Cannot equate block values outside of the game")
        }

        return false
    }

    companion object {
        val CODEC: MapCodec<BlockValue> = RecordCodecBuilder.mapCodec { instance ->
            instance.group(
                Registries.BLOCK.getCodec().fieldOf("value").forGetter<BlockValue>(BlockValue::value)
            ).apply(instance, ::BlockValue)
        }
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + valueType.hashCode()
        return result
    }
}
